### PR TITLE
Feat/eoa_deploys_contract

### LINKED ISSUE #491: feat: add deploy support in EOA / Kakarot
# Feature Request

Deploy is done in EVM by sending a raw transaction to address "empty string" 
And calldata := bytecode! 

EOA needs to handle this case. 
Idea: by branching ifs on the `to` field inside the transaction (e.g. if to if empty string -> kakarot.deploy, else kakarot.execute_at_address)

### PR DESCRIPTION
**Pull request by TotalPizza**
Time spent on this PR: 0.5d

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Account contracts have to be deployed by interaction with kakarot and can't be deployed via the EOA. 

Resolves #491 

## What is the new behavior?

- Sending a transaction (legacy or default) with the "to" address set to "" will result in a contract deployment instead of a standard transaction.
- An event containing the new contracts EVM address is emitted on contract creation

### CODE REVIEW

File: src/kakarot/accounts/eoa/library.cairo [hunk 1]
```diff
@@ -122,18 +124,34 @@ namespace ExternallyOwnedAccount {
...
         let (current_tx_calldata: felt*) = alloc();
-        assert [current_tx_calldata] = destination;
-        assert [current_tx_calldata + 1] = amount;
-        assert [current_tx_calldata + 2] = gas_limit;
-        assert [current_tx_calldata + 3] = payload_len;
-        memcpy(current_tx_calldata + 4, payload, payload_len);
+        local offset;
+        local selector;
+        // If destination is 0, we are deploying a contract
+        if (destination == FALSE) {
```
Comment (TotalPizza) [inline src/kakarot/accounts/eoa/library.cairo:132]: Alternatively I could move the logic into their own functions:
```
if  (destination == FALSE) {
    deploy_contract();
}else{
    execute_contract_bytecode();
}
```
But imo this is still fine.

File: tests/unit/src/kakarot/accounts/eoa/test_library.py [hunk 2]
```diff
@@ -45,17 +45,31 @@ async def test_execute_should_make_all_calls_and_return_concat_results(
...
-                transaction["value"],
-                transaction["gas"],
-                len(
-                    bytes.fromhex(transaction.get("data", "").replace("0x", ""))
-                ),  # calldata_len
-                *list(
-                    bytes.fromhex(transaction.get("data", "").replace("0x", ""))
-                ),  # calldata
-            ]
+            # Execute contract bytecode
```
Comment (TotalPizza) [inline tests/unit/src/kakarot/accounts/eoa/test_library.py:48]: @ClementWalter  kinda hijacked your test here.
I didn't think that this deserved its own one, as we're still only testing whether transactions are decoded and forwarded to kakarot correctly.

**Comment by TotalPizza** [comment]
I could extend this PR to include the new contracts evm address being determined using the EVMs nonce instead of the global kakarot salt.
(But it's probably better as it's own PR, as each account_contract now also requires an individual nonce that it uses when CREATE or CREATE2 is being executed)

File: src/kakarot/accounts/library.cairo [hunk 3]
```diff
@@ -27,6 +27,10 @@ from kakarot.constants import (
 )
 from kakarot.interfaces.interfaces import IAccount
 
+@event
+func evm_contract_deployed(evm_contract_address: felt, starknet_contract_address: felt) {
```
Comment (TotalPizza) [inline src/kakarot/accounts/library.cairo:31]: @Eikix Is there more info we might need for the rpc/block-explorer?
Maybe the evm address that deployed the contract?

File: tests/unit/src/kakarot/accounts/eoa/mock_kakarot.cairo [hunk 4]
```diff
@@ -13,3 +13,13 @@ func execute_at_address{
 ) {
     return (address, value, gas_limit, calldata_len, calldata);
 }
+
+// @dev mock kakarot contract that returns inputs for execute_at_address
```
Comment (ClementWalter) [inline tests/unit/src/kakarot/accounts/eoa/mock_kakarot.cairo:17]: update plz

File: tests/utils/constants.py [hunk 5]
```diff
@@ -74,4 +74,23 @@
...
         "chainId": CHAIN_ID,
     },
+    {
+        "type": 2,
+        "gas": 100_000,
+        "maxFeePerGas": 2_000_000_000,
+        "maxPriorityFeePerGas": 2_000_000_000,
+        "data": "0x616263646566",
+        "nonce": 34,
+        "to": "",
```
Comment (ClementWalter) [inline tests/utils/constants.py:84]: is `"to": ""` the same as `"to": ADDRESS_ZERO`?

**Review by ClementWalter — APPROVED** [review]
lgtm, just few questions

File: tests/utils/constants.py (hunk 5 above)
Comment (TotalPizza) [reply tests/utils/constants.py:84]: I remember @Eikix saying during the hackathon that `""` was what they would be sending via the RPC, which is why I used it. But yes, `""` and `ZERO_ADDRESS` both result in `to` being **0** inside the contract.

File: src/utils/eth_transaction.cairo [hunk 6]
```diff
@@ -245,6 +245,11 @@ namespace EthTransaction {
         let (gas_limit, destination, amount, payload_len, payload, tx_hash, v, r, s) = decode(
             tx_data_len, tx_data
         );
+        if (destination == FALSE){
+            with_attr error_message("ExternallyOwnedAccount: account creations are not payable") {
```
Comment (Eikix) [inline src/utils/eth_transaction.cairo:249]: I think they are? granted the constructor is payable as well (i think)

File: src/kakarot/kakarot.cairo [hunk 7]
```diff
@@ -167,17 +167,18 @@ func set_native_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_chec
...
+// @param gas_limit: maximum amount of gas the constructor can use
 // @param bytes_len: the constructor + contract bytecode lenght
 // @param bytes: the constructor + contract bytecode
 // @return evm_contract_address The evm address that is mapped to the newly deployed starknet contract address
 // @return starknet_contract_address The newly deployed starknet contract address
 @external
 func deploy_contract_account{
     syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr, bitwise_ptr: BitwiseBuiltin*
-}(bytecode_len: felt, bytecode: felt*) -> (
+}(gas_limit: felt, bytecode_len: felt, bytecode: felt*) -> (
```
Comment (Eikix) [inline src/kakarot/kakarot.cairo:178]: I don't think we need gas_limit as a param, we decided in the community call to deprecate the use of gas_limit in our logic.
I think a pragmatic approach would be to remove this most recent addition of gas_limit inside the function

File: tests/utils/constants.py (hunk 5 above)
Comment (Eikix) [reply tests/utils/constants.py:84]: very cool !

File: src/kakarot/accounts/eoa/library.cairo (hunk 1 above)
Comment (Eikix) [reply src/kakarot/accounts/eoa/library.cairo:132]: still fine imo as well

File: src/kakarot/accounts/library.cairo (hunk 3 above)
Comment (Eikix) [reply src/kakarot/accounts/library.cairo:31]: I'm not sure yet, does emitting an event pose a problem to the app?

I recall @ClementWalter saying it was ruining the event parsing in his tests

**Review by Eikix — CHANGES_REQUESTED** [review]
very cool PR! let's merge it asap and try to deploy via remix

File: src/utils/eth_transaction.cairo (hunk 6 above)
Comment (TotalPizza) [reply src/utils/eth_transaction.cairo:249]: Just noticed, you're 100% correct.
Fixing that asap.

File: src/kakarot/kakarot.cairo (hunk 7 above)
Comment (TotalPizza) [reply src/kakarot/kakarot.cairo:178]: Okidoki. Removing it.

File: src/kakarot/accounts/library.cairo (hunk 3 above)
Comment (TotalPizza) [reply src/kakarot/accounts/library.cairo:31]: Oh, I thought it just got removed because it wasn't being used.

Comment (ClementWalter) [reply src/kakarot/accounts/library.cairo:31]: Actually, there is not real problem it was only for events parsing I’m the web3.pu lib.

Comment (Eikix) [reply src/kakarot/accounts/library.cairo:31]: We can leave it there!

**Review by Eikix — APPROVED** [review]
lgtm

File: src/kakarot/kakarot.cairo [hunk 8]
```diff
@@ -167,6 +167,7 @@ func set_native_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_chec
 }
 
 // @notice Deploy a new contract account and execute constructor
+// @param gas_limit: maximum amount of gas the constructor can use
```
Comment (ClementWalter) [inline src/kakarot/kakarot.cairo:170]: this need to be removed then

File: src/kakarot/library.cairo [hunk 9]
```diff
@@ -164,6 +164,7 @@ namespace Kakarot {
     // @notice deploy contract account
     // @dev First deploy a contract_account with no bytecode, then run the calldata as bytecode with the new address,
     //      then set the bytecode with the result of the initial run
+    // @param gas_limit: maximum amount of gas the constructor can use
```
Comment (ClementWalter) [inline src/kakarot/library.cairo:167]: remove then plz

**Review by ClementWalter — DISMISSED** [review]

**Review by ClementWalter — APPROVED** [review]

File: src/kakarot/kakarot.cairo (hunk 8 above)
Comment (TotalPizza) [reply src/kakarot/kakarot.cairo]: whooops,
removed!

**Pull request merged**